// @liskhq/lisk-transaction-pool/dist-node/queue_checkers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lisk_cryptography_1 = require("@liskhq/lisk-cryptography");
exports.checkTransactionPropertyForValues = (values, propertyName) => (transaction) => {
    if (propertyName === 'recipientId') {
        return transaction.asset.recipientId &&
            typeof transaction.asset.recipientId === 'string'
            ? values.includes(transaction.asset.recipientId)
            : false;
    }
    if (propertyName === 'senderId') {
        return values.includes(lisk_cryptography_1.getAddressFromPublicKey(transaction.senderPublicKey));
    }
    return values.includes(transaction[propertyName]);
};
exports.returnTrueUntilLimit = (limit) => {
    let current = 0;
    return _ => current++ < limit;
};
exports.checkTransactionForExpiry = () => {
    const timeNow = new Date();
    return (transaction) => transaction.isExpired(timeNow);
};
exports.checkTransactionForSenderPublicKey = (transactions) => {
    const senderProperty = 'senderPublicKey';
    const senderPublicKeys = transactions.map(transaction => transaction[senderProperty]);
    return exports.checkTransactionPropertyForValues(senderPublicKeys, senderProperty);
};
exports.checkTransactionForId = (transactions) => {
    const idProperty = 'id';
    const ids = transactions.map(transaction => transaction.id);
    return exports.checkTransactionPropertyForValues(ids, idProperty);
};
exports.checkTransactionForSenderIdWithRecipientIds = (transactions) => {
    const recipientProperty = 'recipientId';
    const senderIdProperty = 'senderId';
    const recipients = transactions
        .map(transaction => transaction.asset[recipientProperty])
        .filter(id => id !== undefined);
    return exports.checkTransactionPropertyForValues(recipients, senderIdProperty);
};
exports.checkTransactionForTypes = (transactions) => {
    const typeProperty = 'type';
    const types = transactions.map((transaction) => transaction[typeProperty]);
    return exports.checkTransactionPropertyForValues(types, typeProperty);
};
