// @liskhq/lisk-transactions/dist-node/utils/sign_and_validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cryptography = require("@liskhq/lisk-cryptography");
const errors_1 = require("../errors");
exports.validateSignature = (publicKey, signature, transactionBytes, id) => {
    const transactionHash = cryptography.hash(transactionBytes);
    const valid = cryptography.verifyData(transactionHash, signature, publicKey);
    return {
        valid,
        error: !valid
            ? new errors_1.TransactionError(`Failed to validate signature ${signature}`, id, '.signature')
            : undefined,
    };
};
exports.signaturesAreUnique = (signatures) => {
    const uniqueSignatures = [...new Set(signatures)];
    if (uniqueSignatures.length !== signatures.length) {
        return false;
    }
    return true;
};
exports.checkPublicKeySignatureUniqueness = (publicKeys, signatures, transactionBytes, id) => {
    const checkedPublicKeys = new Set();
    const validSignatures = new Set();
    publicKeys.forEach(publicKey => {
        signatures.forEach((signature) => {
            if (checkedPublicKeys.has(publicKey) || validSignatures.has(signature)) {
                return;
            }
            const { valid: signatureValid } = exports.validateSignature(publicKey, signature, transactionBytes, id);
            if (signatureValid) {
                checkedPublicKeys.add(publicKey);
                validSignatures.add(signature);
            }
        });
    });
    return validSignatures;
};
exports.validateMultisignatures = (publicKeys, signatures, minimumValidations, transactionBytes, id) => {
    if (!exports.signaturesAreUnique(signatures)) {
        return {
            valid: false,
            errors: [
                new errors_1.TransactionError('Encountered duplicate signature in transaction', id, '.signatures'),
            ],
        };
    }
    const validSignatures = exports.checkPublicKeySignatureUniqueness(publicKeys, signatures, transactionBytes, id);
    const invalidTransactionSignatures = signatures.filter(signature => !validSignatures.has(signature));
    if (signatures.length < minimumValidations) {
        return {
            valid: false,
            errors: [
                new errors_1.TransactionPendingError(`Missing signatures`, id, '.signatures'),
            ],
        };
    }
    return {
        valid: validSignatures.size >= minimumValidations &&
            invalidTransactionSignatures.length === 0,
        errors: invalidTransactionSignatures.length > 0
            ? invalidTransactionSignatures.map(signature => new errors_1.TransactionError(`Failed to validate signature ${signature}`, id, '.signatures'))
            : [],
    };
};
