// @liskhq/lisk-p2p/dist-node/peer/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const socketClusterClient = require("socketcluster-client");
const constants_1 = require("../constants");
const errors_1 = require("../errors");
const events_2 = require("../events");
const p2p_request_1 = require("../p2p_request");
const utils_1 = require("../utils");
exports.socketErrorStatusCodes = {
    ...socketClusterClient.SCClientSocket.errorStatuses,
    1000: 'Intentionally disconnected',
};
const RATE_NORMALIZATION_FACTOR = 1000;
var ConnectionState;
(function (ConnectionState) {
    ConnectionState["CONNECTING"] = "connecting";
    ConnectionState["OPEN"] = "open";
    ConnectionState["CLOSED"] = "closed";
})(ConnectionState = exports.ConnectionState || (exports.ConnectionState = {}));
class Peer extends events_1.EventEmitter {
    constructor(peerInfo, peerConfig) {
        super();
        this._peerConfig = peerConfig;
        this._peerInfo = this._initializeInternalState(peerInfo);
        this._rateInterval = this._peerConfig.rateCalculationInterval;
        this._counterResetInterval = setInterval(() => {
            this._resetCounters();
        }, this._rateInterval);
        this._productivityResetInterval = setInterval(() => {
            this._resetProductivity();
        }, constants_1.DEFAULT_PRODUCTIVITY_RESET_INTERVAL);
        this._serverNodeInfo = peerConfig.serverNodeInfo;
        this._handleRawRPC = (packet, respond) => {
            let rawRequest;
            try {
                rawRequest = utils_1.validateRPCRequest(packet);
            }
            catch (error) {
                respond(error);
                this.emit(events_2.EVENT_INVALID_REQUEST_RECEIVED, {
                    packet,
                    peerId: this._peerInfo.peerId,
                });
                return;
            }
            if (events_2.PROTOCOL_EVENTS_TO_RATE_LIMIT.has(rawRequest.procedure) &&
                this._peerInfo.internalState.rpcCounter.has(rawRequest.procedure)) {
                this._updateRPCCounter(rawRequest);
                return;
            }
            this._updateRPCCounter(rawRequest);
            const rate = this._getRPCRate(rawRequest);
            const request = new p2p_request_1.P2PRequest({
                procedure: rawRequest.procedure,
                data: rawRequest.data,
                id: this.peerInfo.peerId,
                rate,
                productivity: this.internalState.productivity,
            }, respond);
            this.emit(events_2.EVENT_REQUEST_RECEIVED, request);
        };
        this._handleWSMessage = () => {
            this._peerInfo.internalState.wsMessageCount += 1;
        };
        this._handleRawMessage = (packet) => {
            let message;
            try {
                message = utils_1.validateProtocolMessage(packet);
            }
            catch (error) {
                this.emit(events_2.EVENT_INVALID_MESSAGE_RECEIVED, {
                    packet,
                    peerId: this._peerInfo.peerId,
                });
                return;
            }
            this._updateMessageCounter(message);
            const rate = this._getMessageRate(message);
            const messageWithRateInfo = {
                ...message,
                peerId: this._peerInfo.peerId,
                rate,
            };
            if (message.event === events_2.REMOTE_EVENT_POST_NODE_INFO) {
                this._handleUpdatePeerInfo(message);
            }
            this.emit(events_2.EVENT_MESSAGE_RECEIVED, messageWithRateInfo);
        };
    }
    get id() {
        return this._peerInfo.peerId;
    }
    get ipAddress() {
        return this._peerInfo.ipAddress;
    }
    get wsPort() {
        return this._peerInfo.wsPort;
    }
    get internalState() {
        return this.peerInfo.internalState;
    }
    get state() {
        const state = this._socket
            ? this._socket.state === this._socket.OPEN
                ? ConnectionState.OPEN
                : ConnectionState.CLOSED
            : ConnectionState.CLOSED;
        return state;
    }
    updateInternalState(internalState) {
        this._peerInfo = {
            ...this._peerInfo,
            internalState,
        };
    }
    get peerInfo() {
        return this._peerInfo;
    }
    _initializeInternalState(peerInfo) {
        return peerInfo.internalState
            ? peerInfo
            : {
                ...peerInfo,
                internalState: utils_1.assignInternalInfo(peerInfo, this._peerConfig.secret),
            };
    }
    updatePeerInfo(newPeerInfo) {
        this._peerInfo = {
            sharedState: newPeerInfo.sharedState,
            internalState: this._peerInfo.internalState,
            ipAddress: this.ipAddress,
            wsPort: this.wsPort,
            peerId: this.id,
        };
    }
    connect() {
        if (!this._socket) {
            throw new Error('Peer socket does not exist');
        }
    }
    disconnect(code = constants_1.INTENTIONAL_DISCONNECT_CODE, reason) {
        clearInterval(this._counterResetInterval);
        clearInterval(this._productivityResetInterval);
        if (this._socket) {
            this._socket.destroy(code, reason);
        }
    }
    send(packet) {
        if (!this._socket) {
            throw new Error('Peer socket does not exist');
        }
        this._socket.emit(events_2.REMOTE_SC_EVENT_MESSAGE, {
            event: packet.event,
            data: packet.data,
        });
    }
    async request(packet) {
        return new Promise((resolve, reject) => {
            if (!this._socket) {
                throw new Error('Peer socket does not exist');
            }
            this._socket.emit(events_2.REMOTE_SC_EVENT_RPC_REQUEST, {
                procedure: packet.procedure,
                data: packet.data,
            }, (error, responseData) => {
                if (error) {
                    reject(error);
                    return;
                }
                if (responseData) {
                    resolve(responseData);
                    return;
                }
                reject(new errors_1.RPCResponseError(`Failed to handle response for procedure ${packet.procedure}`, `${this.ipAddress}:${this.wsPort}`));
            });
        });
    }
    async fetchPeers() {
        try {
            const response = await this.request({
                procedure: events_2.REMOTE_EVENT_RPC_GET_PEERS_LIST,
            });
            const validatedPeers = utils_1.validatePeerInfoList(response.data, this._peerConfig.maxPeerDiscoveryResponseLength, this._peerConfig.maxPeerInfoSize);
            return validatedPeers.map(peerInfo => ({
                ...peerInfo,
                sourceAddress: this.ipAddress,
            }));
        }
        catch (error) {
            if (error instanceof errors_1.InvalidPeerInfoError ||
                error instanceof errors_1.InvalidPeerInfoListError) {
                this.applyPenalty(constants_1.INVALID_PEER_LIST_PENALTY);
            }
            this.emit(events_2.EVENT_FAILED_TO_FETCH_PEERS, error);
            throw new errors_1.RPCResponseError('Failed to fetch peer list of peer', this.ipAddress);
        }
    }
    async discoverPeers() {
        const discoveredPeerInfoList = await this.fetchPeers();
        discoveredPeerInfoList.forEach(peerInfo => {
            this.emit(events_2.EVENT_DISCOVERED_PEER, peerInfo);
        });
        return discoveredPeerInfoList;
    }
    async fetchAndUpdateStatus() {
        let response;
        try {
            response = await this.request({
                procedure: events_2.REMOTE_EVENT_RPC_GET_NODE_INFO,
            });
        }
        catch (error) {
            this.emit(events_2.EVENT_FAILED_TO_FETCH_PEER_INFO, error);
            throw new errors_1.RPCResponseError('Failed to fetch peer info of peer', `${this.ipAddress}:${this.wsPort}`);
        }
        try {
            this._updateFromProtocolPeerInfo(response.data);
        }
        catch (error) {
            this.emit(events_2.EVENT_FAILED_PEER_INFO_UPDATE, error);
            if (error instanceof errors_1.InvalidPeerInfoError) {
                this.applyPenalty(constants_1.INVALID_PEER_INFO_PENALTY);
            }
            throw new errors_1.RPCResponseError('Failed to update peer info of peer due to validation of peer compatibility', `${this.ipAddress}:${this.wsPort}`);
        }
        this.emit(events_2.EVENT_UPDATED_PEER_INFO, this._peerInfo);
        return this._peerInfo;
    }
    applyPenalty(penalty) {
        this.peerInfo.internalState.reputation -= penalty;
        if (this.internalState.reputation <= 0) {
            this._banPeer();
        }
    }
    _resetCounters() {
        this._peerInfo.internalState.wsMessageRate =
            (this.peerInfo.internalState.wsMessageCount * RATE_NORMALIZATION_FACTOR) /
                this._rateInterval;
        this._peerInfo.internalState.wsMessageCount = 0;
        if (this.peerInfo.internalState.wsMessageRate >
            this._peerConfig.wsMaxMessageRate) {
            const messageRateExceedCoefficient = Math.floor(this.peerInfo.internalState.wsMessageRate /
                this._peerConfig.wsMaxMessageRate);
            const penaltyRateMultiplier = messageRateExceedCoefficient > 1 ? messageRateExceedCoefficient : 1;
            this.applyPenalty(this._peerConfig.wsMaxMessageRatePenalty * penaltyRateMultiplier);
        }
        this._peerInfo.internalState.rpcRates = new Map([...this.internalState.rpcCounter.entries()].map(([key, value]) => {
            const rate = value / this._rateInterval;
            if (events_2.PROTOCOL_EVENTS_TO_RATE_LIMIT.has(key) && value > 1) {
                this.applyPenalty(this._peerConfig.wsMaxMessageRatePenalty);
            }
            return [key, rate];
        }));
        this._peerInfo.internalState.rpcCounter = new Map();
        this._peerInfo.internalState.messageRates = new Map([...this.internalState.messageCounter.entries()].map(([key, value]) => {
            const rate = value / this._rateInterval;
            return [key, rate];
        }));
        this._peerInfo.internalState.messageCounter = new Map();
        return;
    }
    _resetProductivity() {
        if (this.peerInfo.internalState.productivity.lastResponded <
            Date.now() - constants_1.DEFAULT_PRODUCTIVITY_RESET_INTERVAL) {
            this._peerInfo.internalState.productivity = { ...constants_1.DEFAULT_PRODUCTIVITY };
        }
    }
    _updateFromProtocolPeerInfo(rawPeerInfo) {
        if (!this._serverNodeInfo) {
            throw new Error('Missing server node info.');
        }
        const peerInfo = utils_1.validatePeerInfo(utils_1.sanitizeIncomingPeerInfo({
            ...rawPeerInfo,
            ipAddress: this.ipAddress,
            wsPort: this.wsPort,
        }), this._peerConfig.maxPeerInfoSize);
        const result = utils_1.validatePeerCompatibility(peerInfo, this._serverNodeInfo);
        if (!result.success && result.error) {
            throw new Error(`${result.error} : ${peerInfo.ipAddress}:${peerInfo.wsPort}`);
        }
        this.updatePeerInfo(peerInfo);
    }
    _handleUpdatePeerInfo(message) {
        try {
            this._updateFromProtocolPeerInfo(message.data);
        }
        catch (error) {
            if (error instanceof errors_1.InvalidPeerInfoError) {
                this.applyPenalty(constants_1.INVALID_PEER_INFO_PENALTY);
            }
            this.emit(events_2.EVENT_FAILED_PEER_INFO_UPDATE, error);
            return;
        }
        this.emit(events_2.EVENT_UPDATED_PEER_INFO, this.peerInfo);
    }
    _banPeer() {
        this.emit(events_2.EVENT_BAN_PEER, this.id);
        this.disconnect(constants_1.FORBIDDEN_CONNECTION, constants_1.FORBIDDEN_CONNECTION_REASON);
    }
    _updateRPCCounter(packet) {
        const key = packet.procedure;
        const count = (this.internalState.rpcCounter.get(key) || 0) + 1;
        this.peerInfo.internalState.rpcCounter.set(key, count);
    }
    _getRPCRate(packet) {
        const rate = this.peerInfo.internalState.rpcRates.get(packet.procedure) || 0;
        return rate * RATE_NORMALIZATION_FACTOR;
    }
    _updateMessageCounter(packet) {
        const key = packet.event;
        const count = (this.internalState.messageCounter.get(key) || 0) + 1;
        this.peerInfo.internalState.messageCounter.set(key, count);
    }
    _getMessageRate(packet) {
        const rate = this.internalState.messageRates.get(packet.event) || 0;
        return rate * RATE_NORMALIZATION_FACTOR;
    }
}
exports.Peer = Peer;
