// @liskhq/lisk-p2p/dist-node/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class PeerInboundHandshakeError extends Error {
    constructor(message, statusCode, remoteAddress, handshakeURL) {
        super(message);
        this.name = 'PeerInboundHandshakeError';
        this.statusCode = statusCode;
        this.remoteAddress = remoteAddress;
        this.handshakeURL = handshakeURL;
    }
}
exports.PeerInboundHandshakeError = PeerInboundHandshakeError;
class RPCResponseError extends Error {
    constructor(message, peerId) {
        super(message);
        this.name = 'RPCResponseError';
        this.peerId = peerId;
    }
}
exports.RPCResponseError = RPCResponseError;
class InvalidRPCResponseError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidRPCResponseError';
    }
}
exports.InvalidRPCResponseError = InvalidRPCResponseError;
class RPCResponseAlreadySentError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ResponseAlreadySentError';
    }
}
exports.RPCResponseAlreadySentError = RPCResponseAlreadySentError;
class ExistingPeerError extends Error {
    constructor(peerInfo) {
        super('Peer already exists');
        this.name = 'ExistingPeerError';
        this.peerInfo = peerInfo;
    }
}
exports.ExistingPeerError = ExistingPeerError;
class InvalidNodeInfoError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidNodeInfoError';
    }
}
exports.InvalidNodeInfoError = InvalidNodeInfoError;
class InvalidPeerInfoError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidPeerInfoError';
    }
}
exports.InvalidPeerInfoError = InvalidPeerInfoError;
class InvalidPeerInfoListError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidPeerInfoListError';
    }
}
exports.InvalidPeerInfoListError = InvalidPeerInfoListError;
class RequestFailError extends Error {
    constructor(message, response, peerId, peerVersion) {
        super(message);
        this.name = 'RequestFailError';
        this.response = response || new Error(message);
        this.peerId = peerId || '';
        this.peerVersion = peerVersion || '';
        this.message = peerId
            ? `${this.message}: Peer Id: ${this.peerId}: Peer Version: ${this.peerVersion}`
            : message;
    }
}
exports.RequestFailError = RequestFailError;
class SendFailError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SendFailError';
    }
}
exports.SendFailError = SendFailError;
class InvalidRPCRequestError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidRPCRequestError';
    }
}
exports.InvalidRPCRequestError = InvalidRPCRequestError;
class InvalidProtocolMessageError extends Error {
    constructor(message) {
        super(message);
        this.name = 'InvalidProtocolMessageError';
    }
}
exports.InvalidProtocolMessageError = InvalidProtocolMessageError;
