// @liskhq/lisk-transactions/dist-node/register_multisignature_account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lisk_validator_1 = require("@liskhq/lisk-validator");
const _12_multisignature_transaction_1 = require("./12_multisignature_transaction");
const constants_1 = require("./constants");
const utils_1 = require("./utils");
const validateInputs = ({ keysgroup, lifetime, minimum, networkIdentifier, }) => {
    if (!lisk_validator_1.isValidInteger(lifetime) ||
        lifetime < constants_1.MULTISIGNATURE_MIN_LIFETIME ||
        lifetime > constants_1.MULTISIGNATURE_MAX_LIFETIME) {
        throw new Error(`Please provide a valid lifetime value. Expected integer between ${constants_1.MULTISIGNATURE_MIN_LIFETIME} and ${constants_1.MULTISIGNATURE_MAX_LIFETIME}.`);
    }
    if (!lisk_validator_1.isValidInteger(minimum) ||
        minimum < constants_1.MULTISIGNATURE_MIN_KEYSGROUP ||
        minimum > constants_1.MULTISIGNATURE_MAX_KEYSGROUP) {
        throw new Error(`Please provide a valid minimum value. Expected integer between ${constants_1.MULTISIGNATURE_MIN_KEYSGROUP} and ${constants_1.MULTISIGNATURE_MAX_KEYSGROUP}.`);
    }
    if (keysgroup.length < minimum) {
        throw new Error('Minimum number of signatures is larger than the number of keys in the keysgroup.');
    }
    lisk_validator_1.validateKeysgroup(keysgroup, constants_1.MULTISIGNATURE_MIN_KEYSGROUP, constants_1.MULTISIGNATURE_MAX_KEYSGROUP);
    lisk_validator_1.validateNetworkIdentifier(networkIdentifier);
};
exports.registerMultisignature = (inputs) => {
    validateInputs(inputs);
    const { keysgroup, lifetime, minimum, passphrase, secondPassphrase, networkIdentifier, } = inputs;
    const plusPrependedKeysgroup = utils_1.prependPlusToPublicKeys(keysgroup);
    const keygroupFees = plusPrependedKeysgroup.length + 1;
    const transaction = {
        ...utils_1.createBaseTransaction(inputs),
        type: 12,
        fee: (constants_1.MULTISIGNATURE_FEE * keygroupFees).toString(),
        asset: {
            min: minimum,
            lifetime,
            keysgroup: plusPrependedKeysgroup,
        },
        networkIdentifier,
    };
    if (!passphrase) {
        return transaction;
    }
    const multisignatureTransaction = new _12_multisignature_transaction_1.MultisignatureTransaction(transaction);
    multisignatureTransaction.sign(passphrase, secondPassphrase);
    return multisignatureTransaction.toJSON();
};
